Streaming requests

legionth opened the pull request.

legionth committed:
Streaming requests to the callback function

legionth commented:
Ping @clue

clue commented on src/HttpServer.php:
```diff
@@ -71,7 +71,7 @@ public function handleConnection(ConnectionInterface $connection)
                     $request = RingCentral\Psr7\parse_request($fullHeader);
                 } catch (\Exception $ex) {
-                    $that->sendResponse(new Response(400), $connection);
+                    $connection->write(RingCentral\Psr7\str(new Response(400)));
```
Motivation?

clue commented on examples/streamReceivingServer.php:
```diff
@@ -0,0 +1,44 @@
+            $bodyContent .= $chunk;
+        });
+
+            $body->on('end', function () use (&$bodyContent, $resolve) {
```
Indentation is a bit off here. Also, I wonder why this is different from the example in the README?

legionth replied:
In the newest commit these examples are equal.

clue commented on src/HttpServer.php:
```diff
@@ -103,42 +103,36 @@ public function handleConnection(ConnectionInterface $connection)
-        $bodyBuffer = '';
-        $that = $this;
-
-        $protection = new CloseProtectionStream($connection);
```
Is this intentional? I wonder why this protection isn't needed anymore?

legionth replied:
No it wasn't :hospital:

clue commented on src/HttpServer.php:
```diff
@@ -103,42 +103,36 @@ public function handleConnection(ConnectionInterface $connection)
+            $bodyStream = new HttpBodyStream($connection);
+            $request = $request->withBody($bodyStream);
             $this->handleRequest($connection, $request);
+            $bodyStream->emit('end', array());
```
Not sure if this is tested, but could the connection keep emitting `data` events here?

legionth replied:
You are right. The  comment from below showed that data would be emitted.

clue commented on tests/HttpServerTest.php:
```diff
@@ -428,15 +443,21 @@ public function testBiggerBodyThanContentLengthWillBeCutted()
         $callback = function(RequestInterface $request) {
-            return new Response();
+            $promise = new Promise(function ($resolve, $reject) use ($request) {
+                $request->getBody()->on('end', function () use ($resolve) {
```
See also above. We likely need to explicitly ensure no `data` event will be emitted here. May also affect other tests afaict?

legionth replied:
I added an exception throw to some tests if any data will be emitted.

clue reviewed (changes requested):
Can't await this feature :-) :+1:

legionth committed:
Extend Tests

legionth committed:
Use same example in README and examples folder

legionth committed:
Add PHP 5.3 support

legionth commented:
Ping @clue

clue commented on src/HttpBodyStream.php:
```diff
@@ -64,6 +64,11 @@ public function close()
+        $this->input->removeListener('data', array($this, 'handleData'));
+        $this->input->removeListener('end', array($this, 'handleEnd'));
+        $this->input->removeListener('error', array($this, 'handleError'));
+        $this->input->removeListener('close', array($this, 'close'));
```
Is this needed? Closing the input should remove all listeners?

legionth replied:
Sure. When a connection is the input stream. The data will be still handled.

clue replied:
Not sure who's responsible here. The input stream will be closed 2 lines down. Shouldn't the input stop emitting events after it is closed?

If not, isn't the input misbehaving here? (and this should be fixed instead?)

legionth replied:
The input will stop I just thought it would be better to remove the listeners. But ok I'll remove them.

clue commented on src/HttpServer.php:
```diff
@@ -255,4 +251,10 @@ function () use ($connection) {
     }
+
+    /** @internal */
+    public function sendResponse(ResponseInterface $response, ConnectionInterface $connection)
```
Is this related? Should this be a separate PR? Also looks like some tests were missing here?

clue replied:
This should probably also be tested for invalid requests?

clue commented on README.md:
```diff
@@ -216,6 +217,48 @@ If you use the `HttpBodyStream` the whole transfer will be [chunked encoded](htt
+
+This is just an example you can use a [BufferedSink](https://github.com/reactphp/stream) from the `reactphp/stream` to avoid these lines of code.
+
+Check out the `examples` folder how your server could look like.
```
The docs LGTM :+1: 

What if I happen have built on top of the existing version (non-streaming)? It looks like the upcoming version *only* supports streaming. Any recommendations/docs on how I should upgrade?

legionth committed:
Rework of examples and READEME for streaming requests

legionth committed:
Revert remove of listeners

legionth commented:
Updated README and adjusted examples. I also thought about to remove the `Response` object as return value of the callback function and only allow promise. What do you think @clue ?

clue reviewed (approved):
Nice, LGTM! :+1: :tada: 

I guess this makes sense from a consistency perspective, but I kind of dislike how every handler now *has* to wait for the `end` event. I'm not sure this is something that can be addresses now, so I'll leave this up to you to decide :+1:

clue commented:
> remove the Response object as return value

Given the current flow, I guess this would make sense, but I kind of miss how easy the previous interface was...

I'm undecided currently, what do you think about this? What if I want to build a really simple API and don't want to care about request bodies at all?

legionth commented:
I think this is good Idea will add something before a new version. I think about something.

legionth committed:
Adding tests to ensure connection will be closed for further data

legionth committed:
Add better description from simple request handling to request streaming

legionth commented:
Ping @clue. Added 2 commits.

clue commented on src/HttpServer.php:
```diff
@@ -104,42 +104,37 @@ public function handleConnection(ConnectionInterface $connection)
+            $bodyStream = new HttpBodyStream($protection);
+            $request = $request->withBody($bodyStream);
             $this->handleRequest($connection, $request);
+            $bodyStream->close();
```
Does it make sense to close the stream *before* or *after* it is passed to the callback function?

A closed stream will not emit an `end` event, however, it's state could be checked via `isReadable()`.

I'm undecided, what do you think?

clue reviewed (approved):
Nice! The updated documentation makes this much much easier to grasp!

legionth committed:
Updated README to explain the body of the request object
